moz_security/sandboxes/heka/analysis/moz_security_hh.lua

```diff
@@ -0,0 +1,187 @@
+seen for a given identification field (e.g., the IP address in an nginx log). Can be utilized to trend
+heavy usage from specific identifiers.
+
+The plugin by default outputs collected data as TSV; if the send_tigerblood configuration option
+is enabled the plugin will instead output data as Tigerblood violation messages (for consumption by
+for example a Tigerblood output module).
+
+## Sample Configuration
+```lua
+filename = "moz_security_heavy_hitters.lua"
+message_matcher = "Logger == 'input.nginx'"
+ticker_interval = 60
+preserve_data = true
+
+id_field = "Fields[remote_addr]" -- field to use as the identifier
+-- id_field_capture = ",? *([^,]+)$",  -- optional e.g. extract the last entry in a comma delimited list
+
+-- alist_items = 10000 -- optional, defaults to 10000 (maximum number of heavy hitter IDs to track)
+-- send_tigerblood = false -- optional, if true plugin will generate Tigerblood violation messages
+-- violation_type = "fxa:heavy_hitter_ip" -- required in violations mode, Tigerblood violation type
+
+sample_min_id = 5000 -- Minimum distinct identifers before a sample will be calculated
+sample_max_id = 10000 -- Maximum identifiers to use for sample calculation within a window
+sample_min_ev = 50000 -- Minimum number of events sampler must consume before calculation
+sample_window = 60 -- Sample window size
+sample_ticks = 2500 -- Recalculate sample every sample_ticks events
+threshold_cap = 10 -- Threshold will be calculated average + (calculated average * cap)
+-- cms_epsilon = 1 / 10000 -- optional CMS value for epsilon
+-- cms_delta = 0.0001 -- optional CMS value for delta
+```
+--]]
+
+require "string"
+require "table"
+
+local ostime    = require "os".time
+
+local sample_min_id = read_config("sample_min_id") or error("sample_min_id must be configured")
+local sample_max_id = read_config("sample_max_id") or error("sample_max_id must be configured")
+local sample_min_ev = read_config("sample_min_ev") or error("sample_min_ev must be configured")
+local sample_window = read_config("sample_window") or error("sample_window must be configured")
+local sample_ticks  = read_config("sample_ticks") or error ("sample_ticks must be configured")
+local threshold_cap = read_config("threshold_cap") or error("threshold_cap must be configured")
+local id_field      = read_config("id_field") or error("id_field must be configured")
+local id_fieldc     = read_config("id_field_capture")
+local cms_epsilon   = read_config("cms_epsilon") or 1 / 10000
+local cms_delta     = read_config("cms_delta") or 0.0001
+local alist_items   = read_config("alist_items") or 10000
+
+local tbsend
+local violation_type
+local send_tigerblood = read_config("send_tigerblood")
+if send_tigerblood then
+    tbsend = require "heka.tigerblood".send
+    violation_type = read_config("violation_type") or error("violation_type must be configured")
+end
+
+local cms = require "streaming_algorithms.cm_sketch".new(cms_epsilon, cms_delta)
+
+local sampler = {}
```

Comment:
As discussed the sampler adds a lot of overhead and complexity and should be factored out in future versions.

See logging-cep-test mtrinkala_mozilla.com.hh1.lua for comparison.